### PR TITLE
resource_control: support dynamic calibrate resource

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/38825

Problem Summary:
The maximum RU estimated by this PR is based on an actual and running workload by user. And user can set the time point.

Similar to https://github.com/pingcap/tidb/pull/42165, we only consider TiDB CPU or TiKV CPU as bottleneck. Also,  the resource consuming is linear co-related with each other.
For each metrics sampling point, the PR calculates the RU quota at each point in time using RU statistics, tidb CPU statistics, and tikv statistics. Then removes the 10% maximum and 10% minimum, then calculates the average. In addition, if CPU resource utilization is low at some point in time, it will not be included in the calculation

And ref https://github.com/tikv/pd/issues/6298, update pd client.
### Check List

Tests 

- [x] Unit test
- [x] Manual test (add detailed scripts or steps below)
image

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
calibrate resource support dynamic calibrate for user-actual workload with specific time point.
```

### CODE REVIEW

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go:97]: const (... )

File: parser/ast/misc.go
Comment (glorv) [inline parser/ast/misc.go:3798]: Why depending on this flag?

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go]: Maybe can uniform case of words in each statement.

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go]: remove useless log

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go]: remove useless log

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go]: Perhaps these check infos duplicate the infos in getFunc.

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go:111]: Can we extract max and min time to const?

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:119]: I'd like to replace this if-else with something like following:
```
var start, end, dur *ptr
for _, op := range e.optionList[0] {
  ...
}
if duration == nil { ...default for duration}
if start == nil { ...default-for-start }
if end == nil { ...default-for-end }

validate_start_end_duration()

...rest logics
```

And Since The static and dynamic branch have little common logic with each other, please wrap both of them with a separate function to avoid too long if..else.. block

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go]: useless log

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:207]: What if `tikvCPUs[idx] == 0` here, I think check `tikvCPUs[idx]/totalKVCPUQuota` as the cpu usage percentage is a more ergonomic way

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go:120]: `e.optionList[1].Tp != ast.CalibrateEndTime && e.optionList[1].Tp != ast.CalibrateDuration`
I'm not sure why the same parameter would have && judgment twice

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:212]: I think if one of the two cpu usage is greater than the lowUsageThreshold, we should keep it. Maybe there are cluster topologies that tidb cpu quota >> tikv cpu 
 quota or vice verse, then no samples can be valid here.

File: parser/ast/misc.go
Comment (HuSharp) [inline parser/ast/misc.go:3810]: Maybe can return error directly, because there is same check at `if err := option.Restore(ctx); err != nil {`

File: parser/misc.go
Comment (HuSharp) [inline parser/misc.go:326]: Maybe need Camel-Case

File: parser/parser.y
Comment (HuSharp) [inline parser/parser.y:390]: ditto

File: parser/ast/misc.go
Comment (CabinfeverB) [reply parser/ast/misc.go:3798]: No need. Should also remove it in `ResourceGroupOption`.

File: executor/calibrate_resource.go
Comment (CabinfeverB) [reply executor/calibrate_resource.go:120]: Because we're trying to determine if it's wrong. The expression to assert truth is `e.optionList[1].Tp == ast.CalibrateEndTime || e.optionList[1].Tp == ast.CalibrateDuration`

File: executor/calibrate_resource.go
Comment (CabinfeverB) [reply executor/calibrate_resource.go:212]: How about add `valuableUsageThreshold`?  If one of the two cpu usage is greater than the `valuableUsageThreshold`, we can accept it

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go:180]: Maybe can put this check in `dynamicCalibrate`?

File: parser/parser.y
Comment (HuSharp) [inline parser/parser.y:720]: I noticed there is a `startTS` located below, do we need to use `startTS`?

Comment (CabinfeverB) [reply parser/parser.y:720]: IMO, START_TIME is better

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go:181]:
```suggestion
	startTs, endTs, err := e.checkDynamicCalibrateOptions()
	if err != nil {
		return err
	}
	startTime := startTs.In(e.ctx.GetSessionVars().Location()).Format("2006-01-02 15:04:05")
	endTime := endTs.In(e.ctx.GetSessionVars().Location()).Format("2006-01-02 15:04:05")
```

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go:205]: maybe can  reduce the initial number of rows
```suggestion
	lowCount, tidbCPULowCount, tikvCPULowCount := 0, 0, 0
```
BTW, `tikvCPULowCOunt` has a extra big `O`

File: executor/calibrate_resource.go
Comment (HuSharp) [inline executor/calibrate_resource.go:211]: ditto

File: parser/parser.y
Comment (CabinfeverB) [reply parser/parser.y]: update it

File: executor/calibrate_resource.go
Comment (CabinfeverB) [reply executor/calibrate_resource.go]: update it

**Review by HuSharp — COMMENTED** [review]
small advice, rest LGTM!

File: parser/parser_test.go
Comment (glorv) [inline parser/parser_test.go:1077]: I suppose this case should return false.
BTW, Please also add a test for:
```
calibrate resource START_TIME = '2023-04-01 13:00:00', END_TIME = '2023-04-01 16:00:00', duration = "20m"
calibrate resource workload = oltp_read_write, duration = "20m"
```

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:117]:
```suggestion
			ts, err = staleread.CalculateAsOfTsExpr(e.ctx, op.Ts)
```

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:123]:
```suggestion
			ts, err = staleread.CalculateAsOfTsExpr(e.ctx, op.Ts)
```

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:129]:
```suggestion
			dur, err = duration.ParseDuration(op.StrValue)
```

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:111]:
```suggestion
func (e *calibrateResourceExec) parseCalibrateDuration() (startTime time.Time, endTime time.Time, err error) {
```

File: executor/calibrate_resource.go
Comment (glorv) [reply executor/calibrate_resource.go:117]: Seems the logic of this function is not well covered by unit test, please add some cases

**Comment by HuSharp** [comment]
Maybe need to change Error Code
MySQL [(none)]>  CALIBRATE RESOURCE START_TIME '2023-04-18 08:01:00' DURATION '9m';
ERROR 1105 (HY000): the duration of calibration is too short, should be greater than 10m0s

image

File: executor/calibrate_resource.go
Comment (songrijie) [inline executor/calibrate_resource.go:150]: Suggested msg:
"The duration is too long/short, which could lead to inaccurate output. Please make the duration between s% and s%. "

File: executor/calibrate_resource.go
Comment (songrijie) [inline executor/calibrate_resource.go:244]: Suggest msg
"The workload in selected time window is too low, with which TiDB is unable to reach a capacity estimation. Please select another time window with higher workload, or calibrate resource by hardware instead. "

**Comment by CabinfeverB** [comment]
> 1105

1105 is not specifical for the error  that occurs when the version of TiDB does not match with that of TiKV

File: parser/parser.y
Comment (glorv) [reply parser/parser.y:737]: I prefer start_time too. StartTs represent the pd txn timestamp in this context, but the start_time is a real datetime, better not to mix them.

Comment (HuSharp) [reply parser/parser.y:737]: Makes sense

File: executor/calibrate_resource.go
Comment (glorv) [inline executor/calibrate_resource.go:95]: Please add comments for these constants

**Review by HuSharp — APPROVED** [review]
LGTM

File: go.mod
Comment (nolouch) [inline go.mod:284]: why replace it?

Comment (CabinfeverB) [reply go.mod:284]: there are two PRs will be merged in /pd/client, I want to wait until they are merged

Comment (nolouch) [reply go.mod:284]: better use a separate pr to update it.

Comment (CabinfeverB) [reply go.mod]: update

**Review by nolouch — APPROVED** [review]

**Pull request merged**
